### PR TITLE
Elasticache updates

### PR DESCRIPTION
**Pull request by ocampocj**
Worth mentioning that the finalsnapshot trigger also causes any tags to be copied from the cluster to the snapshot.  No additional code was needed for that.

### CODE REVIEW

File: c7n/resources/elasticache.py [hunk 1]
```diff
@@ -88,14 +119,16 @@ def process(self, clusters):
...
             else:
                 clusters_to_delete.append(cluster)
-
+        # added if statement to handle differences in parameters if snapshot is skipped
         for cluster in clusters_to_delete:
             params = {'CacheClusterId': cluster['CacheClusterId']}
             if _cluster_eligible_for_snapshot(cluster) and not skip:
                 params['FinalSnapshotIdentifier'] = snapshot_identifier(
                     'Final', cluster['CacheClusterId'])
+                self.log.info("Taking final snapshot of %s" %cluster['CacheClusterId'])
```
Comment (kapilt) [inline c7n/resources/elasticache.py]: debug level logging on this, pls

File: c7n/resources/elasticache.py [hunk 2]
```diff
@@ -88,14 +119,16 @@ def process(self, clusters):
...
+            else:
+                self.log.info("Skipping final snapshot of %s" %cluster['CacheClusterId'])
```
Comment (kapilt) [inline c7n/resources/elasticache.py]: this too

File: c7n/resources/elasticache.py [hunk 3]
```diff
@@ -239,11 +272,13 @@ def _elasticache_cluster_tags(
     def process_tags(cluster):
         client = local_session(session_factory).client('elasticache')
         arn = generator(cluster[model.id])
-        tag_list = retry(
-            client.list_tags_for_resource,
-            ResourceName=arn)['TagList']
-        cluster['Tags'] = tag_list or []
-        return cluster
+        # added if statement to ensure snapshot is available in order to list tags
+        if cluster['CacheClusterStatus'] == 'available':
```
Comment (kapilt) [inline c7n/resources/elasticache.py]: if not available return, ie. dedent block

File: c7n/resources/elasticache.py [hunk 4]
```diff
@@ -254,18 +289,24 @@ def _elasticache_snapshot_tags(
...
+    # added if statement to ensure snapshot is available in order to list tags
     def process_tags(snapshot):
         client = local_session(session_factory).client('elasticache')
         arn = generator(snapshot[model.id])
-        tag_list = retry(
-            client.list_tags_for_resource,
-            ResourceName=arn)['TagList']
-        snapshot['Tags'] = tag_list or []
-        return snapshot
+        if snapshot['SnapshotStatus'] == 'available':
```
Comment (kapilt) [inline c7n/resources/elasticache.py]: ditto, if not available return, dedent block

File: c7n/resources/elasticache.py [hunk 5]
```diff
@@ -254,18 +289,24 @@ def _elasticache_snapshot_tags(
...
     with executor_factory(max_workers=2) as w:
         return list(w.map(process_tags, snapshots))
 
 
 def _cluster_eligible_for_snapshot(cluster):
-    return cluster['Engine'] != 'memcached'
+    # added regex search to filter unsupported cachenode types
+    return (
+        cluster['Engine'] != 'memcached' and not 
+        re.search('cache.t', cluster['CacheNodeType'])
```
Comment (kapilt) [inline c7n/resources/elasticache.py]: can you compile this as a top level module constant, avoids regex compilation per cluster.

File: tests/test_elasticache.py [hunk 6]
```diff
@@ -39,6 +39,37 @@ def test_elasticache_cluster_simple_filter(self):
...
+                 'value': 'memcached'}],
+            'actions': [
+                {'type': 'mark-for-op', 'days': 30,
+                'op': 'delete'}]},
+            session_factory=session_factory)
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
+        
+    def test_elasticache_cluster_unmark(self):
+        session_factory = self.replay_flight_data('test_elasticache_cluster_unmark')
```
Comment (kapilt) [inline tests/test_elasticache.py]: please describe the test setup when there's no inline setup with the code.

**Comment by kapilt** [comment]
a couple of minors, but in general looks good, thanks. bonus for a test on non available cluster to validate the additional snapshot/tag logic for future maintenance.

**Comment by ocampocj** [comment]
Thanks, Kapil.  I think I've covered all your comments in my latest commits.

File: c7n/resources/elasticache.py [hunk 7]
```diff
@@ -33,7 +35,10 @@
 
 filters = FilterRegistry('elasticache.filters')
 actions = ActionRegistry('elasticache.actions')
+#registered marked-for-op filter
+filters.register('marked-for-op', tags.TagActionFilter) 
 
+ttype = re.compile('cache.t')
```
Comment (kapilt) [inline c7n/resources/elasticache.py]: generally for global variables, that are referenced from inside functions, best practice is to UPPER_CASE them so they stand out. 

else looks good, thanks

**Comment by kapilt** [comment]
lgtm

**Pull request merged**